https://leetcode.com/problems/perfect-number/

# Brute Force Solution
- O(n)

In [9]:
class Solution:
    def checkPerfectNumber(self, num: int) -> bool:
        divisors = set()
        
        for i in range(1, num):
            if num % i == 0:
                divisors.add(i)
                
        return sum(divisors) == num
    
    def perfectNumsInRange(self, upper_bound):
        perfect_nums = []
        for i in range(upper_bound):
            if self.checkPerfectNumber(i):
                perfect_nums.append(i)
                
        return perfect_nums

In [10]:
s = Solution()
%time s.perfectNumsInRange(100)

CPU times: user 428 µs, sys: 3 µs, total: 431 µs
Wall time: 434 µs


[0, 6, 28]

In [12]:
s = Solution()
%time s.perfectNumsInRange(1000)

CPU times: user 27.6 ms, sys: 1.03 ms, total: 28.7 ms
Wall time: 27.8 ms


[0, 6, 28, 496]

In [14]:
s = Solution()
%time s.perfectNumsInRange(10000)

CPU times: user 2.41 s, sys: 6.22 ms, total: 2.42 s
Wall time: 2.42 s


[0, 6, 28, 496, 8128]

### Perfect Numbers in range (1 - 100000)
- we can see that the runtime starts to grow at an increasing rate as our upper bound scales
- it only took 27.8 ms for the first 1000 numbers
- it took 4 minutes 12 seconds to scan the first 100000 numbers!
- we need a more efficient algorithm to be able to find larger perfect numbers

In [15]:
s = Solution()
%time s.perfectNumsInRange(100000)

CPU times: user 4min 10s, sys: 545 ms, total: 4min 11s
Wall time: 4min 12s


[0, 6, 28, 496, 8128]

# Optimal Solution
- O(sqrt(n))

In [18]:
class Solution2:
    def checkPerfectNumber(self, num: int) -> bool:
        if num == 1:
            return False
        res = 1
        for i in range(2,int(num**0.5)+1):
            if num%i == 0:
                res += i + num//i
        return res == num
    
    def perfectNumsInRange(self, upper_bound):
        perfect_nums = []
        for i in range(upper_bound):
            if self.checkPerfectNumber(i):
                perfect_nums.append(i)
                
        return perfect_nums
            

In [20]:
s = Solution2()

%time s.perfectNumsInRange(100000)

CPU times: user 869 ms, sys: 4.66 ms, total: 874 ms
Wall time: 877 ms


[6, 28, 496, 8128]

In [24]:
s = Solution2()

%time s.perfectNumsInRange(1000000)

CPU times: user 32.6 s, sys: 113 ms, total: 32.7 s
Wall time: 32.9 s


[6, 28, 496, 8128]